plat/common/lcpu.c

```diff
@@ -48,3 +191,447 @@ void ukplat_lcpu_halt_to(__nsec until)
+			 * caller explicitly put the executing CPU in, we still
+			 * ignore it but need to skip the parameters.
+			 */
+			if (lcpuidx)
+				argi++;
+
+			continue;
+		}
+
+retry:
+		old = ukarch_load_n(&lcpu->state);
+
+		/* We ignore CPUs that are already started */
+		if (unlikely(old != LCPU_STATE_OFFLINE)) {
+			uk_pr_warn("Failed to start CPU 0x%lx: not offline\n",
+				   lcpu->id);
+
+			/* Skip CPU and its arguments*/
+			argi++;
+			continue;
+		}
+
+		/* Try to acquire the CPU for initialization. If another thread
+		 * was faster, we will return to the state comparison and
+		 * report that the CPU is not offline.
+		 */
+		if (ukarch_compare_exchange_sync((int *)&lcpu->state, old,
+						 new) != new)
+			goto retry;
+
+		UK_ASSERT(lcpu->state == LCPU_STATE_INIT);
+
+		/* Setup startup arguments.
+		 * Since we are ignoring the executing CPU, we must keep a
+		 * separate counter to index the arguments.
+		 */
+		lcpu->s_args.entry = (entry && entry[argi]) ?
+			entry[argi] : lcpu_entry_default;
+		lcpu->s_args.stackp = sp[argi];
+
+		/* Ensure that the startup arguments have been written back
+		 * before issuing the startup call
+		 */
+		wmb();
+
+		rc = lcpu_arch_start(lcpu, flags);
+		if (unlikely(rc)) {
+			lcpu->state = LCPU_STATE_HALTED;
+			lcpu->error_code = rc;
+
+			/* There is a serious problem. Stop here. The caller
+			 * can skip the CPU by using the value of *num.
+			 */
+			break;
+		}
+
+		/* Move to next argument */
+		argi++;
+	}
+
+#ifdef LCPU_ARCH_MULTI_PHASE_STARTUP
+	/* At this point, i has been set to the number of successfully
+	 * started CPUs. So if there has been an error, we won't touch
+	 * any CPUs not started.
+	 */
+	rc2 = lcpu_arch_post_start(lcpuidx, &i);
+	if (unlikely(rc2)) {
+		if (num) {
+			UK_ASSERT(i <= *num);
+			*num = i;
+		}
+
+		/* Return the first error */
+		return (rc) ? rc : rc2;
+	}
+#endif /* LCPU_ARCH_MULTI_PHASE_STARTUP */
+
+	UK_ASSERT(num == NULL || *num == i);
+	return rc;
+}
+
+static inline int lcpu_transition_safe(struct lcpu *lcpu, int incr)
+{
+	int old, new;
+
+	/* Transition the CPU to a different busy level. The CPU could not be
+	 * online or fall into a halted state at any moment, we thus cannot
+	 * just atomically in-/decrement the state. Otherwise, we might corrupt
+	 * the non-online state.
+	 */
+	do {
+		old = ukarch_load_n(&lcpu->state);
+
+		/* We must not change the state if the CPU is not online */
+		if (!lcpu_state_is_online(old))
+			return 0;
+
+		UK_ASSERT(old <= INT_MAX - incr);
+		UK_ASSERT(old >= INT_MIN + incr);
+		new = old + incr;
+
+		UK_ASSERT(lcpu_state_is_online(new));
+	} while (ukarch_compare_exchange_sync((int *)&lcpu->state, old,
+					      new) != new);
+
+	return 1;
+}
+
+int ukplat_lcpu_run(const __lcpuidx lcpuidx[], unsigned int *num,
+		    const struct ukplat_lcpu_func *fn, unsigned long flags)
+{
+	__lcpuid this_cpu_id = ukplat_lcpu_id();
+	struct lcpu *lcpu;
+	unsigned int n, i;
+	int rc, rc2 __maybe_unused;
```

Comment:
`rc2` is not used in this function and it wasn't caught because of `__maybe_unused`.